aws/resource_aws_budgets_budget_test.go

```diff
@@ -0,0 +1,408 @@
+	if end, _ := time.Parse("2006-01-02_15:04", config.TimePeriodEnd); *timePeriod.End != end {
+		return fmt.Errorf("TimePeriodEnd not set properly '%v' should be '%v'", *timePeriod.End, end)
+	}
+
+	if start, _ := time.Parse("2006-01-02_15:04", config.TimePeriodStart); *timePeriod.Start != start {
+		return fmt.Errorf("TimePeriodStart not set properly '%v' should be '%v'", *timePeriod.Start, start)
+	}
+
+	return nil
+}
+
+func checkBudgetCostTypes(config budgetTestConfig, costTypes budgets.CostTypes) error {
+	if strconv.FormatBool(*costTypes.IncludeCredit) != config.IncludeCredit {
+		return fmt.Errorf("IncludeCredit not set properly '%v' should be '%v'", *costTypes.IncludeCredit, config.IncludeCredit)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeOtherSubscription) != config.IncludeOtherSubscription {
+		return fmt.Errorf("IncludeOtherSubscription not set properly '%v' should be '%v'", *costTypes.IncludeOtherSubscription, config.IncludeOtherSubscription)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeRecurring) != config.IncludeRecurring {
+		return fmt.Errorf("IncludeRecurring not set properly  '%v' should be '%v'", *costTypes.IncludeRecurring, config.IncludeRecurring)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeRefund) != config.IncludeRefund {
+		return fmt.Errorf("IncludeRefund not set properly '%v' should be '%v'", *costTypes.IncludeRefund, config.IncludeRefund)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeSubscription) != config.IncludeSubscription {
+		return fmt.Errorf("IncludeSubscription not set properly '%v' should be '%v'", *costTypes.IncludeSubscription, config.IncludeSubscription)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeSupport) != config.IncludeSupport {
+		return fmt.Errorf("IncludeSupport not set properly '%v' should be '%v'", *costTypes.IncludeSupport, config.IncludeSupport)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeTax) != config.IncludeTax {
+		return fmt.Errorf("IncludeTax not set properly '%v' should be '%v'", *costTypes.IncludeTax, config.IncludeTax)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeUpfront) != config.IncludeUpfront {
+		return fmt.Errorf("IncludeUpfront not set properly '%v' should be '%v'", *costTypes.IncludeUpfront, config.IncludeUpfront)
+	}
+
+	if strconv.FormatBool(*costTypes.UseBlended) != config.UseBlended {
+		return fmt.Errorf("UseBlended not set properly '%v' should be '%v'", *costTypes.UseBlended, config.UseBlended)
+	}
+
+	return nil
+}
+
+func testCheckBudgetDestroy(budgetName string, provider *schema.Provider) error {
+	meta := provider.Meta()
+	client := meta.(*AWSClient).budgetconn
+	accountID := meta.(*AWSClient).accountid
+	_, err := client.DescribeBudget(&budgets.DescribeBudgetInput{
+		BudgetName: &budgetName,
+		AccountId:  &accountID,
+	})
+	if !isAWSErr(err, budgets.ErrCodeNotFoundException, "") {
+		return fmt.Errorf("Budget '%s' was not deleted properly", budgetName)
+	}
+
+	return nil
+}
+
+type budgetTestConfig struct {
```

Comment:
If we could remove this custom configuration indirection here, that would be wonderful. Being more verbose with the configurations and the attribute testing (looking inside the `*budget.Budget` from the API directly or even creating a `&budget.Budget{}` for reflect.DeepEqual) makes the testing easier to understand.

Reply:
agreed, this struct always kind of bothered me. intent was to flatten the config so it was more manageable, but to your point it likely just ends up adding complexity.